src/java/org/apache/cassandra/journal/Journal.java

```diff
@@ -659,6 +657,47 @@ void selectSegmentToFlush(Collection<ActiveSegment<K, V>> into)
         segments().selectActive(currentSegment.descriptor.timestamp, into);
     }
 
+    ActiveSegment<K, V> oldestActiveSegment()
+    {
+        ActiveSegment<K, V> current = currentSegment;
+        if (current == null)
+            return null;
+
+        ActiveSegment<K, V> oldest = segments().oldest();
+        if (oldest == null || oldest.descriptor.timestamp > current.descriptor.timestamp)
+            return current;
+
+        return oldest;
+    }
+
+    ActiveSegment<K, V> currentActiveSegment()
+    {
+        return currentSegment;
+    }
+
+    ActiveSegment<K, V> getActiveSegment(long timestamp)
+    {
+        // we can race with segment addition to the segments() collection, with a new segment appearing in currentSegment first
+        // since we are most likely to be requesting the currentSegment anyway, we resolve this case by checking currentSegment first
+        // and resort to the segments() collection only if we do not match
+        ActiveSegment<K, V> currentSegment = this.currentSegment;
+        long currentSegmentTimestamp = currentSegment.descriptor.timestamp;
+        if (timestamp == currentSegmentTimestamp)
+        {
+            return currentSegment;
+        }
+        else if (timestamp > currentSegmentTimestamp)
+        {
+            throw new IllegalArgumentException("Requested a newer timestamp " + timestamp + " than the current active segment " + currentSegmentTimestamp);
+        }
+        else
+        {
+            Segment<K, V> segment = segments().get(timestamp);
+            Invariants.checkState(segment != null, "Segment %d expected to be found, but neither current segment %d nor in active segments", timestamp, currentSegmentTimestamp);
+            return segment.asActive();
```

Comment:
Given the way it's currently used, I imagine it should never be a static segment, but maybe still ensure that we are only querying for active segments? Maybe rename `get` to `getActive()` and add an `isActive()` check there, and the more specific return type so no casting is needed here?